### PR TITLE
add grant function on AuthService

### LINKED ISSUE #15713: Add grant function on AuthService (abstract service)

### PR DESCRIPTION
**Pull request by mahmut-gundogdu**
### Description

Resolves #15713 (write the related issue number if available)

TODO: Describe what this PR has changed, add screenshot or animated GIF if available, write if it is a breaking change, and how to fix the breaking changes for existing applications if so.

### Checklist

- [x] I fully tested it as developer / designer and created unit / integration tests
- [ ] I documented it (or no need to document or I will create a separate documentation issue)

### How to test it?
no need to test on abp repo.  it just abstraction

### CODE REVIEW

File: npm/ng-packs/packages/core/src/lib/abstracts/abstract.model.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export interface AbstractModel {
```
Comment (muhammedaltug) [inline npm/ng-packs/packages/core/src/lib/abstracts/abstract.model.ts]: I think we can rename it to OAuthResponse

File: npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts [hunk 2]
```diff
@@ -1,7 +1,9 @@
-import { Injectable } from '@angular/core';
-import { Params } from '@angular/router';
-import { Observable, of } from 'rxjs';
-import { LoginParams } from '../models/auth';
+import {HttpHeaders} from '@angular/common/http';
```
Comment (muhammedaltug) [inline npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts]: please reformat

File: npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts [hunk 3]
```diff
@@ -10,7 +12,8 @@ import { LoginParams } from '../models/auth';
   providedIn: 'root',
 })
 export class AuthService implements IAuthService {
-  constructor() {}
+  constructor() {
```
Comment (muhammedaltug) [inline npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts]: we can remove empty constructor

File: npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts [hunk 4]
```diff
@@ -31,7 +34,8 @@ export class AuthService implements IAuthService {
     return of(undefined);
   }
 
-  navigateToLogin(queryParams?: Params): void {}
+  navigateToLogin(queryParams?: Params): void {
```
Comment (muhammedaltug) [inline npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts]: please reformat file

File: npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts [hunk 5]
```diff
@@ -56,4 +65,6 @@ export interface IAuthService {
   navigateToLogin(queryParams?: Params): void;
 
   login(params: LoginParams): Observable<any>;
+
+  loginUsingGrant(grantType: string, parameters: object, headers?: HttpHeaders): Promise<AbstractModel>;
```
Comment (muhammedaltug) [inline npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts]: this method can be optional

File: npm/ng-packs/packages/oauth/src/lib/services/oauth.service.ts [hunk 6]
```diff
@@ -1,24 +1,27 @@
...
-import { filter, map, switchMap, take, tap } from 'rxjs/operators';
-import { IAuthService, LoginParams } from '@abp/ng.core';
-import { AuthFlowStrategy } from '../strategies';
-import { EnvironmentService } from '@abp/ng.core';
-import { AUTH_FLOW_STRATEGY } from '../tokens/auth-flow-strategy';
-import { AuthConfig, OAuthService } from "angular-oauth2-oidc";
+import {Injectable, Injector} from '@angular/core';
+import {Params} from '@angular/router';
+import {from, Observable, lastValueFrom} from 'rxjs';
+import {filter, map, switchMap, take, tap} from 'rxjs/operators';
```
Comment (muhammedaltug) [inline npm/ng-packs/packages/oauth/src/lib/services/oauth.service.ts]: please reformat file

File: npm/ng-packs/packages/core/src/lib/abstracts/abstract.model.ts (hunk 1 above)
Comment (mahmut-gundogdu) [reply npm/ng-packs/packages/core/src/lib/abstracts/abstract.model.ts]: The user don't have to use Oauth, so I have Changed AbpAuthResponse

File: npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts (hunk 5 above)
Comment (mahmut-gundogdu) [reply npm/ng-packs/packages/core/src/lib/abstracts/auth.service.ts]: I think we can't mark as optional. That used on Abp commercial. I know the user have not use Oauth but we need that function for impersonate and link account.

**Review by muhammedaltug — APPROVED** [review]

**Pull request merged**